### PR TITLE
`cur_data()` / `cur_data_all()` not simplifying list columns 

### LINKED ISSUE #5901: `cur_data_all()` for rowwise tibble un-nests list variables
When accessing current group in a rowwise tibble, the list column's nested values should probably not be expanded.  If I have a column with tibbles, they are un-nested and this prevents further processing of the tibble as-is.

``` r
library(dplyr, warn.conflicts = FALSE)

f <- function(x) { print(x, n = nrow(x) )}
bla <- tibble(i = 1:2, x = list(tibble(a = 1), tibble(a = 2))) %>% 
  rowwise() %>% 
  summarise(f(cur_data_all()))
#> # A tibble: 1 x 2
#>       i   x$a
#>   <int> <dbl>
#> 1     1     1
#> # A tibble: 1 x 2
#>       i   x$a
#>   <int> <dbl>
#> 1     2     2
```

<sup>Created on 2021-06-03 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

The same code with `group_by(i)` does not un-nest the tibble.

### PR DESCRIPTION
**Pull request by romainfrancois**
closes #5901

``` r
library(dplyr, warn.conflicts = FALSE)

f <- function(x) { print(x, n = nrow(x) )}
bla <- tibble(i = 1:2, x = list(tibble(a = 1), tibble(a = 2))) %>% 
  rowwise() %>% 
  summarise(f(cur_data_all()))
#> # A tibble: 1 x 2
#>       i x               
#>   <int> <list>          
#> 1     1 <tibble [1 × 1]>
#> # A tibble: 1 x 2
#>       i x               
#>   <int> <list>          
#> 1     2 <tibble [1 × 1]>
```

<sup>Created on 2021-09-17 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

### CODE REVIEW

**Comment by romainfrancois** [comment]
However, the data is still simplified as usual when used directly in the expression, or through `across()`: 

``` r
library(dplyr, warn.conflicts = FALSE)
library(tibble)

df <- tibble(x = list(tibble(a = 1), tibble(a = 2))) %>%
  rowwise()

summarise(df, is_tibble(x))
#> # A tibble: 2 x 1
#>   `is_tibble(x)`
#>   <lgl>         
#> 1 TRUE          
#> 2 TRUE
summarise(df, across(x, is_tibble))
#> # A tibble: 2 x 1
#>   x    
#>   <lgl>
#> 1 TRUE 
#> 2 TRUE
```

<sup>Created on 2021-09-17 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

File: NEWS.md [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # dplyr (development version)
 
+* `cur_data()` and `cur_data_all()` not simplifying list columns rowwise data frames (#5901).
```
Comment (hadley) [inline NEWS.md:3]:
```suggestion
* `cur_data()` and `cur_data_all()` don't simplify list columns in rowwise data frames (#5901).
```

File: R/data-mask.R [hunk 2]
```diff
@@ -67,6 +67,14 @@ DataMask <- R6Class("DataMask",
 
     pick = function(vars) {
       cols <- self$current_cols(vars)
+      if (inherits(private$data, "rowwise_df")) {
```
Comment (hadley) [inline R/data-mask.R:70]: Can you explain why we have to re-wrap in a list? (As opposed to not unwrapping somewhere?)

Comment (romainfrancois) [reply R/data-mask.R:70]: It happens in this line: 

https://github.com/tidyverse/dplyr/blob/bb84360805084c348b4616857688d6260b139c08/src/chop.cpp#L20

```cpp
void dplyr_lazy_vec_chop_grouped(SEXP chops_env, SEXP rows, SEXP data, bool rowwise) {
  SEXP names = PROTECT(Rf_getAttrib(data, R_NamesSymbol));
  R_xlen_t n = XLENGTH(data);

  const SEXP* p_data = VECTOR_PTR_RO(data);
  const SEXP* p_names = STRING_PTR_RO(names);
  for (R_xlen_t i = 0; i < n; i++) {
    SEXP prom = PROTECT(Rf_allocSExp(PROMSXP));
    SET_PRENV(prom, R_EmptyEnv);
    SEXP column = p_data[i];
    if (rowwise && vctrs::vec_is_list(column) && Rf_length(column) > 0) {
      SET_PRCODE(prom, column);
    } else {
      SET_PRCODE(prom, Rf_lang3(dplyr::functions::vec_chop, column, rows));
    }
    SET_PRVALUE(prom, R_UnboundValue);

    Rf_defineVar(rlang::str_as_symbol(p_names[i]), prom, chops_env);
    UNPROTECT(1);
  }

  UNPROTECT(1);
}
```

When we setup the promise that makes the chops, and this is a rowwise data, we can skip using `vec_chop()` and simply use the list columns as the chops. 

This gives us the rowwise simplification for when we refer to the list column in an expression (not having to `[[1]]`).

Comment (romainfrancois) [reply R/data-mask.R:70]: ... and for non-original columns, i.e. those that are added, they go through the `<DataMask>$add_one()` method: 

```r
add_one = function(name, chunks, result) {
      if (inherits(private$data, "rowwise_df")){
        is_scalar_list <- function(.x) {
          vec_is_list(.x) && length(.x) == 1L
        }
        if (all(map_lgl(chunks, is_scalar_list))) {
          chunks <- map(chunks, `[[`, 1L)
        }
      }

      .Call(`dplyr_mask_add`, private, name, result, chunks)
    },
```

Comment (hadley) [reply R/data-mask.R:70]: Ah makes sense. Thanks!

**Review by hadley — APPROVED** [review]

**Pull request merged**